### ISSUE TITLE
Test context reading 'wrong' application properties file

### ISSUE THREAD

**Issue by pablo-parra** [Microsoft/vscode-java-test#534]
In a Spring Boot sample project using a MySQL db as main data source and H2 as test data source, when launching the test using Java Test Runner extension the application properties file read by the app is the one located in the "main" package instead the one located in the "test" package. Due to that all the test fail as are not finding the expected infrastructure.

The test of the project are working fine in Eclipse.

![application_properties_issue](https://user-images.githubusercontent.com/18135212/50227243-9bce8180-03a5-11e9-8eaa-d7144a1c3e30.png)

**Comment by rutcreate** [comment]
Same problem here. If there is any workaround please share.

For my case, I use DBRider to test database. It will delete tables in database before or after start test case. From this problem cause dev's database tables will be deleted instead of test's database.

**Comment by jdneo** [comment]
@pablo-parra @rutcreate 

It might have a better solution to this issue. For now, you can try the following workaround:

1. Check if the `application.properties` is in `/target/test-classes/`
2. If yes, then open your `launch.test.json` file, in `env` field, add the following setting: `"spring.config.location": "classpath:/target/test-classes/"`
3. Do not forget to set the `default` field as well, otherwise the extension will not using the configuration as the default.

Here is an example:

```json
{
    "run": {
        "default": "todo-app-java-on-azure",
        "items": [
            {
                "name": "todo-app-java-on-azure",
                "projectName": "todo-app-java-on-azure",
                "workingDirectory": "c:\\Users\\sheche\\Desktop\\todo-app-java-on-azure-master",
                "args": [
                ],
                "vmargs": [],
                "env": {
                    "spring.config.location": "classpath:/target/test-classes/"
                },
                "preLaunchTask": ""
            }
        ]
    },
    // ...
}
```

Please let me know if this could help with your problem.

Thanks

**Comment by pablo-parra** [comment]
Hi @jdneo, I have followed the steps you mention and it's working now 👍. 
I guess it is something that could be done automatically by the extension or at least should be mentioned in the "Usage" section of the extension landing page.

Thanks for your help!! :)

**Comment by yaohaizh** [comment]
@jdneo, please document this either on readme or wiki page.

**Comment by pablo-parra** [comment]
I'm doing now further testing and seems that the workaround is fixing the issue only partially. With the referred configuration `"spring.config.location": "classpath:/target/test-classes/"` in the _env_ property we avoid reading the `\main\resources\application.properties` and the default configurations (H2 in my case) are applied. The problem is that now the context is not reading apparently any customized application.properties file so extra properties are ignored. I have checked it creating a new `test` profile (_application-test.properties_ file) and trying to activate the profile with the `spring.profiles.active=test` property in  `\test\resources\application.properties`. When launching the test during the application startup the logs confirm that the profile (and the related application properties file) is ignored: `No active profile set, falling back to default profiles`.

Following the Spring Boot 2.1.0-RELEASE documentation (that is my project framework version) I have tried adding additional properties to _env_ specification such as `"spring.config.additional-location": "classpath:/target/test-classes/"` and  `"spring.config.name": "application.properties"` but with no success

**Comment by jdneo** [comment]
@pablo-parra Thanks, I'll look into it.

**Comment by jdneo** [comment]
Sorry for the late reply.

After some investigation, I found out that it is because the classpath generated from the Java Language Server will put the `target/classes` higher than `target/test-class`, thus failed to load the correct `appllication.properties` for testing.

I'll figure out if there is any solution to correct the order.

**Comment by jdneo** [comment]
Fixed in #629 and will be released within `0.15.0`

**Comment by pmverma** [comment]
Having same problem, waiting for next release!

**Comment by tnatanael** [comment]
Same problem here!

**Comment by EdgarArguelles** [comment]
issue is still happening in version 0.30.0

**Comment by jdneo** [comment]
@EdgarArguelles Could you share a sample project?

**Comment by fvclaus** [comment]
@jdneo This bug is reproducible with Gradle, but not with Maven. I tried the petclinic project with both Gradle and Maven, but was unable to reproduce it with Maven. The project can be found here: https://github.com/fvclaus/vscode-java-test-application-properties.

I have two `application.properties` with the same property `properties-location` which is either `main` or `test`. 
Using Gradle on the command line, the test succeeds:

```
dev@ZBM-WN-61626:~/workspace/java-test-application-properties-gradle$ gradle test

BUILD SUCCESSFUL in 2s
4 actionable tasks: 4 executed
```

If I execute it in vscode, the test fails:

![image](https://user-images.githubusercontent.com/2384738/161312982-05986f1b-ca20-43e0-957a-5f70080fe67b.png)

I believe it is because the order of the main and test classpath is flipped:
```
-cp /home/dev/workspace/java-test-application-properties-gradle/bin/main:/home/dev/workspace/java-test-application-properties-gradle/bin/test...
```

As a workaround, I currently use:

```
		"java.test.config": [
			{
				"name": "local",
				"workingDirectory": "${workspaceFolder}",
				"env": { 
					"spring.config.location": "${workspaceFolder}/src/test/resources/",
				},
			},
		]
```

**Comment by jdneo** [comment]
@fvclaus, it's a known issue that the classpath resolved by Gradle(Buildship) is wrong. In the future we will consider to add a feature that delegate the test execution to Gradle (that will make sure the result is the same as running from command line)

See: https://github.com/microsoft/vscode-java-test/issues/1045#issuecomment-676911725

**Comment by Ortega-Dan** [comment]
Confirming the issue is still present. Looks like it will be fixed with #1045 but the workaround shared [here](https://github.com/microsoft/vscode-java-test/issues/534#issuecomment-1086167714) by @fvclaus works very well. Thanks.

**Comment by jdneo** [comment]
The Gradle Test Delegation (both run and debug) has supported now.

To use this feature, you need to install the latest [Test Runner for Java](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-test) and [Gradle for Java](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-gradle) extension.

To delegate the tests to Gradle, you can set the default testing profile in Testing explorer:
![image](https://github.com/user-attachments/assets/f3e982e0-0340-461c-b6c4-c8c7fe9352f5)
![image](https://github.com/user-attachments/assets/e1a2693c-e6d7-425a-b7d5-6adf878f4ce3)

If you do not want to change the default testing profile, you can trigger an one-time execution via:
![image](https://github.com/user-attachments/assets/1c3f7478-b100-40cf-98c2-6f0357050209)
![image](https://github.com/user-attachments/assets/55fa017c-adb3-4c5a-bc81-55db63834773)
